GRENZELHOFT COLORPLEX RETURNS- adds tailor's swatchbook!

coffeecoffeecoffee opened the pull request:
## About The Pull Request
Once upon a time on a server before Azure was even a glint in anyone's eye, I learned some of my first code adding feather plumes to helmets and updating the then-rancid colors of my favorite role, the Grenzelhoft mercs. While it's not like I earned a history degree or anything, I did a fair amount of research into real-world historical dyes that inspired choices like Cornflower Blue as well as Madroot and Marigold Orange, and I put a large amount of effort into picking out compelling shades that weren't oversaturated and that fit in with the palettes of the game. We inherited these, and for a long time, grubby little Grenzoids like me were happy.

Then, the wheel nation attacked. https://github.com/GeneralPantsuIsBadAtCoding/Azure-Peak/pull/2311

Originally nuking almost all the different preset colors, this PR has been met with majority negative reception (despite adding a bunch of assorted great changes like the dye brush, which I obviously haven't touched.) Later, the "extra shades" were supposedly readded to the dye bin... Except a lot of them _weren't._ As well, most people seem to have complaints centered around the dye wheel, and we were promised "If people abuse it, they should just get banned!" - but now I'm seeing people going around with significantly bright and garish neon-ass vaporwave colors, so we clearly can't expect staff to do _all_ the work - nobody wants to be the color police, and we just don't do things this way if we can help it because it's obviously unhealthy and unrealistic, for example with our _mechanically obligated_ minimum flavortext limit. There are simply always going to be people who, given the option, will not be mature.

More importantly, one of the STATED GOALS of the PR was to have a "Unified color list for everything" — and it doesn't! After revising the PR to try and re-add all the old presets, the old Grenzel colors remained axed, and they were even left with the flavorless basic bitch Royal Family colors like "Red" and "Yellow" and none of the dye bin colors. The majority of us don't have degrees in color theory, don't want to spend half an hour finding perfect thematic shades, and worst of all, it puts an extra menu in-between you and testing another color. Presets are important and clearly influence the aesthetic tones of the game in practice.

<!-- Describe your pull request. Avoid text walls, use concise bullet points for easier readability. Document every change, or this can delay review and even discourage maintainers from merging your PR. -->

## What this does

This *actually* creates a unified color list for everything, with (almost) ALL of the legacy colors from the dye bin _and_ the Grenzelhoft colors, for Grenzel items, the dye bin _AND_ the Royal Family colors!
Also annihilates the color whee- **Actually makes the color wheel a special trait of the tailor and seamstress**
The following legacy color presets have been fiddled with or removed in the interest of cutting the now-massive list down a little and pruning redundant, identical or near-identical shades:

— Royal White and Politely Yuck removed as they're literally identical to Swan's White and Peasant Brown.
— Royal Black removed as it's massively darker than anything that should be allowed, jesus christ. A significantly dark black option is still in, just not THAT black.
— Dark Grey, Royal Yellow (#ffcd43), Red (#a32121) and Orange (#bd6606) removed as they're very, very similar to Blacksteel Grey, Yellow Weld, Mage Red and Royal Orange respectively.
— Woad Blue (the dye bin version, #597fb9) and Royal Purple (#8747b1) were removed for sharing names with the Grenzelhoft Woad Blue (#597fb9) and Royal Purple (#5E4687). The Woad Blue seems just a less distinctive version of the Grenzel Woad Blue and the Royal Purple is similar to Lavender (#865c9c), just oversaturated as hell.
— "Green" changed to "Forest Green" and "Black" changed to "Darkest Night" because they're good shades, and if they're staying, they should get names that don't suck.
— "Royal Azure" removed because it's so fucking neon, jesus christ. It's an eyesore next to literally everything else. We should have a tasteful lapiz lazuli blue as our namesake, in my opinion, if we even want to have an unofficial-official nudge-nudge "Azure Peak" color.

If you loved any of the colors I've left behind, PLEASE speak up and I will reinclude them. I've tried to specifically axe the less compelling boring stuff like "PURPLE" and "BLUE" and shit I think is actually ugly ingame in favor of all the more imaginative shades, because I figured almost nobody is using them, in an effort to take some of the edge off having such an extensive list now.

## Testing Evidence
Here's a Cornflower Blue halfcloak (my favorite from the Grenzel colors!) which was previously only selectable by Grenzel items unless you got the perfect shade on the color wheel:
![dreamseeker_Sen2F2uHkT](https://github.com/user-attachments/assets/78979ca7-885f-4580-b397-afb671592723)

Here's another previously-Grenzel color on a dye bin item, Bog Green on the fancy coat Princes get:
![dreamseeker_SVPc4afxkz](https://github.com/user-attachments/assets/963c7a0f-dba0-4eab-a705-92f5a9c4c83f)

Here's an Orchil winter dress with a Wine Rouge lord's cloak - another color combo with the Grenzel colors that were removed:
![dreamseeker_FyuJgkwZ3R](https://github.com/user-attachments/assets/f62c3888-202f-4a1c-941b-c61d1da661b9)

Here's Orchil and Darkest Night _as royal family colors,_ who previously could only pick things like "yellow", "green" and "black"!
![dreamseeker_k9EFIVCjKy](https://github.com/user-attachments/assets/6fbefa19-43d7-436a-bfc8-1b28dd0f0b14)

You could now have a servant revolt Peasant Brown / Dirt kingdom. YOU COULD HAVE IT ALL - MY EMPIRE OF SHIT.
![dreamseeker_1LOtPW9Emv](https://github.com/user-attachments/assets/741c30db-7296-4b72-a270-3a3108023f4c)

Here's the ABSOLUTE DARKEST color now selectable, Darkest Night, on a full suit of armor. This is significantly stark black and I'm open to removing this, but keep in mind the PREVIOUS system with the wheel allowed even DARKER blacks (even after this was supposedly "fixed",) namely the preset Royal Black (#2b292e). By the way, did you know the crown is dyeable (but really hard to get a good shade on?) Here, it's Maroon.
![dreamseeker_WLBLgkFfOJ](https://github.com/user-attachments/assets/790b1564-ac1b-480d-8286-84f10728974d)

<!-- It's mandatory to test your PR. Provide images, clips or description of how you tested your changes where possible. -->

## Why It's Good For The Game
THESE were the grenzelhoft colors after the last update.

![45326437](https://github.com/user-attachments/assets/ff8004f4-66ed-42bf-8153-63638bffe2d7)
![tiis cannot continue](https://github.com/user-attachments/assets/fdd218c7-7312-4f0d-9b0b-31c427adcab9)

<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial. If you can't, then it probably isn't good for the game in the first place. -->

FreeStylaLT commented on code/modules/roguetown/roguejobs/tailor/dyer.dm:
```diff
@@ -1,16 +1,47 @@
+	"Light Grey" = "#999999",
+	"Dunked in Water" = "#bbbbbb",
+	"Mage Grey" = "#6c6c6c",
+	"Sow's skin"="#CE929F",
+	"Knight's Red"="#933030",
+	"Royal Red"="#8b2323",
+	"Red Ochre" = "#913831",
+	"Maroon" = "#550000",
+	"Mage Red" = "#b8252c",
+	"Royal Orange" = "#df8405",
+	"Madroot Red"="#AD4545",
+	"Marigold Orange"="#E2A844",
+	"Chestnut" = "#613613",
+	"Dirt" = "#7c6d5c",
+	"Peasant Brown" = "#685542",
+	"Russet" = "#7f461b",
+	"Yellow Weld" = "#f4c430",
+	"Yarrow" = "#f0cb76",
+	"Yellow Ochre" = "#cb9d06",
+	"Mage Yellow" = "#c1b144",
+	"Astrata's Yellow"="#FFFD8D",
+	"Olive" = "#98bf64",
+	"Royal Green" = "#264d26",
+	"Forest Green" = "#428138",
+	"Mage Green" = "#759259",
+	"Bog Green"="#375B48",
+	"Seafoam Green"="#49938B",
+	"Royal Teal" = "#249589",
+	"Cornflower Blue"="#749EE8",
+	"Royal Blue" = "#173266",
+	"Woad Blue"="#395480",
+	"Mage Blue" = "#4756d8",
+	"Periwinkle Blue" = "#8f99fb",
+	"Lavender"="#865c9c",
+	"Royal Purple"="#5E4687",
+	"Orchil" = "#66023C",
+	"Wine Rouge"="#752B55",
+	"Royal Magenta" = "#962e5c",
+	"Blacksteel Grey"="#404040",
+	"Darkest Night" = "#414143"
```
I believe this makes it so the Lord / Duke can pick any one of these for their banners.

coffeecoffeecoffee replied:
Yes. Absolutely. This is explicitly stated in the PR and examples of it are shown in the testing. If this isn't desired, I can change this back to the old system for "Lord colors", but what I was able to create with this was cool as fuck.

Stutternov commented:
> > if this is being done can you make it so you can undo the color on helmets (the cloth you attach) because otherwise it's a guessing game of finding the right non-descript color off the list to try and match keep colors. and you don't get a second chance.
> 
> you can do this with accessory recoloring, in the dye bin. I believe?

it recolors the entire helmet not the cloth on it alone

hubertjones commented:
Used dark grey very, very often. Step off my son

radarseas commented:
I think royal black should be kept, solely because its the only closet color you can get to the Rogue cloaks, the Black hoods, Necromancer robes, Necra attire, the guard satchel, or even the black talibard/jupon/surcoat preset color.

Its a nessessity to avoid meta and to keep the evilmaxxing consistancy.

Draganfrukts commented:
The best way I can personally describe this PR is **"One step forward, two steps back"**

### On the added colors,
This is great. I really like this, and I really like that you went out of your way to make them all properly selectable where it actually matters. Great work on this, good job expanding on it. I'm a big fan, love it. Seriously!

### On the color wheel

> I'm seeing people going around with significantly bright and garish neon-ass vaporwave colors

Ahelp it.

> we clearly can't expect staff to do all the work - nobody wants to be the color police

Why complain about people going around with "bright and garish neon-ass vaporwave colors" if "nobody wants to be the color police"? The janis aren't all-knowing, or all-seeing, and they play the game too. **Ahelp it** and it will be solved.

> There are simply always going to be people who, given the option, will not be mature.

Sorry to really hammer the point home here: But have we considered **ahelping** this? 30 seconds of typing after pressing F1 is a lot better than opening a new github branch or, god forbid, bitching about it in #general for the upteenth time. **People who "will not be mature" go unpunished by inaction, not because of staff being overwhelmed or anything remotely similar.**

### The possible solutions moving forward
**Option 1 - More preset colors**
We add even more preset colors. Enough for people to no longer complain that a specific color they want, and _**bloat the list tremendously for no good reason.**
_**There will eventually be infinite colors, because you can't make everyone happy with the presets. Especially people that know what they're doing.**_

**Option 2 - Just keep the damn color wheel.**
If people are going around in bright pink neon armor, that's a player issue that code can provide a *bandaid* fix for, not something you can permanently smudge out. This is why you can, again, take less than half a minute to write up an ahelp and let an admin resolve the matter in another minute.
### **Not to mention that, color wheel or not, people can still just paint their steel armor bright red, or pink, and achieve nearly the exact same result.**

I already see people complaining about the missing colors that weren't re-added /this very instant/
![image](https://github.com/user-attachments/assets/b1915d50-af91-48a7-b2ea-61c8b28bb640)

Removing the color wheel is ultimately a negative for player expression, it's taking a tool away from people that know what they're doing because *certain players* are keen on complaining about things and then doing nothing about it.

> The majority of us don't have degrees in color theory, don't want to spend half an hour finding perfect thematic shades,

You don't need a degree, and you don't need to spend half an hour finding "the perfect shade". The colors added in this PR make this _completely optional_

>  and worst of all, it puts an extra menu in-between you and testing another color.

Nobody _really_ cares about having to click one more button to get to the color menu, do they? :/

Ollanius commented:
I agree with what Dragonfrukts mentioned.

This pull request is very nice, but I disagree with completely removing the color wheel. 
Having that additional layer of customization _is_ good, and allowing people to fine-tune their appearances to _perfectly_ satiate their style-'tisms is great. I personally use the color wheel to apply battle damage to my character's apparel - dirtied cuirasses, headbands fashioned from torched sleeves, gloves that've been soaked knuckle-deep in gore, etcetera.

The only issue, as mentioned, is that it allows people to create explicitly garish dyes. I think that's a consequence inherent to giving people the _freedom of choice_, however. More-over, I don't believe this'd generate an excess strain of extra tickets for the administration: less than what the neon-haired quirkmaxxers generate, at least.

I would say.. simply add a disclaimer to the 'color wheel' option that clearly lets the user know that they *shouldn't* turn themselves into a neon green-and-pink crusader. You could either attach it to the window itself _(similar to how the warning text on the loadout selection menu works)_, or add it as a seperate message that prints when selecting the 'color wheel' option.

Having to spend two seconds to click either option won't be an issue, either, I think. That's from personal experience, though.

I really love the pull request, otherwise! Good work.

Ryan180602 commented:
The color wheel that I added in the PR is limited by `luminescence`. The proc used for it (`color_pick_sanitized_lum`) can be easily overidden to make it base it off `saturation` instead. Admin action/interaction isn't really needed.

I'd also suggest not having *every* colour preset for the flag, maybe. And let the dye barrels have more by using the `extra_colors` list put down. 

Otherwise I'm unbothered. The list is being used as was intended anyhow.


Also `tertiary` dye should be recoloring only the orle - tested it too. Secondary is usually plume. Primary is the helmet. If accessory dyeing colours the *entire* thing that's probably a buuuuug?

coffeecoffeecoffee commented:
> Used dark grey very, very often. Step off my son

Sure, re-adding it! I only ditched it because this is the sheer breadth of different greys across the systems we've got:
![paintdotnet_OHSARu9iSh](https://github.com/user-attachments/assets/02240415-1811-4c0d-9dec-685f6c8c1f5b)
It was like "wow some of these are barely fucking different" but I know how it is to get your hands on _the perfect shade_ (cornflower blue my beloved)

> if this is being done can you make it so you can undo the color on helmets (the cloth you attach) because otherwise it's a guessing game of finding the right non-descript color off the list to try and match keep colors. and you don't get a second chance.

I would LOVE to do this but it might be beyond my coding ability :,) I'm going to try because I would like this too. Anybody that can add this should because god knows I'll fuck it up. I was barely able to add feather plumes in the first place



> I think royal black should be kept, solely because its the only closet color you can get to the Rogue cloaks, the Black hoods, Necromancer robes, Necra attire, the guard satchel, or even the black talibard/jupon/surcoat preset color. (...)

I can consider this but I un-added it because it seemed like you could make inhumanly stark black shit with it that looked like ass. I'm totally open to re-adding it but I think we should be cautious about stark blacks - someone in dev chat was saying we _shouldn't_, so for everyone's consideration: here's the Mortus cloak (vanilla stock black cloak) with Darkest Night:
![dreamseeker_odn6ZaPTWk](https://github.com/user-attachments/assets/01d5fde3-d22a-4c22-b732-2c07bf4ad81a)

And here with Royal Black:
![dreamseeker_nCe2lfBSH2](https://github.com/user-attachments/assets/2a077365-3002-4f74-9faf-7aa0a3c892ac)

It's barely different, one just looks less like it's armor at all. You sure?



> The best way I can personally describe this PR is **"One step forward, two steps back"**
> 
> > I'm seeing people going around with significantly bright and garish neon-ass vaporwave colors
> 
> Ahelp it.
> 
> > we clearly can't expect staff to do all the work - nobody wants to be the color police
> 
> Why complain about people going around with "bright and garish neon-ass vaporwave colors" if "nobody wants to be the color police"? The janis aren't all-knowing, or all-seeing, and they play the game too. **Ahelp it** and it will be solved.
> 
> > There are simply always going to be people who, given the option, will not be mature.
> 
> Sorry to really hammer the point home here: But have we considered **ahelping** this? (...)

>More-over, I don't believe this'd generate an excess strain of extra tickets for the administration: less than what the neon-haired quirkmaxxers generate, at least.

A staff member has literally said they'd rather not deal with this:
> > All for annihilating the color wheel tbh cause I really DO not want to have to punish someone in bright neon pink cause they've got too many notes as is.

It creates work for admins and room for abuse, which - normally, if it's a trade-off for useful expression, is totally fine IMO! - but here, it's for hyperspecific shades that usually don't fit the game, and a menu that's annoying. This will be and is being abused with how accessible it was made. There may be a compromise that I'll get to below - but on _this_ topic, I don't think allowing anything and everything in the name of "player expression" for a handful of people who can be mature with it, when the majority are degrading the roleplay of the game, is a net benefit or responsible at all.

Generally, we don't leave this in the hands of players because staffing is an immensely difficult and busy job already. We don't unlock the flavortext limit "for the people who know what they're doing" because ultimately it just leads to lazy characters. The same seems obviously true here.

>Not to mention that, color wheel or not, people can still just paint their steel armor bright red, or pink, and achieve nearly the exact same result.

This is the entire point of the color presets and was one of the chief concerns when I was originally picking out the shades; that they should be desaturated and fit with the game's palettes. They will never be _as_ bright or immensely saturated a yellow, red, pink etc. as something grabbed off the color wheel; for instance, here's Sow's Pink.
![67358](https://github.com/user-attachments/assets/2b363781-6cbe-429a-a225-8bfb5a479b4b)

> > The majority of us don't have degrees in color theory, don't want to spend half an hour finding perfect thematic shades,
> 
> You don't need a degree, and you don't need to spend half an hour finding "the perfect shade". The colors added in this PR make this _completely optional_

_This_ is an argument I'm much more amenable to. Now that I have the shades _I_ want, I don't care _too_ much if the color wheel's just sitting there on the side, even if I've seen that it sucks for the roleplay on the server. That said, an idea I saw brought up in dev chat that's very interesting is _maybe the color wheel could be exclusive to seamstresses and tailors._ I feel like this would be a super valid compromise which also gives purpose to a role that's currently pretty much "just for flavor", and actually makes roleplay around color; giving them a real market and a unique skill this way sounds cool to me, and being able to get brighter colors through a seamstress makes genuine sense.

Better if we can get a saturation limit, which is beyond my current abilities to code. Someone who's able to do that should re-add the color wheel, but let's not add it without it and then "leave that for later / someone else" while leaving the game and the people roleplaying in it in a worse state.

> > and worst of all, it puts an extra menu in-between you and testing another color.
> Nobody _really_ cares about having to click one more button to get to the color menu, do they? :/

> Having to spend two seconds to click either option won't be an issue, either, I think. That's from personal experience, though.

It was so annoying to my friends and I that I opened up Dreammaker and coded again It is indescribably fucking annoying when looking through the colors, I'm going to be real with you. But that's not... The _worst_ thing in the world nor my chief concern. I could deal with that if it didn't also seem _just bad for the game._

coffeecoffeecoffee commented:
> The color wheel that I added in the PR is limited by `luminescence`. The proc used for it (`color_pick_sanitized_lum`) can be easily overidden to make it base it off `saturation` instead. Admin action/interaction isn't really needed.
> 
> I'd also suggest not having _every_ colour preset for the flag, maybe. And let the dye barrels have more by using the `extra_colors` list put down.
> 
> Otherwise I'm unbothered. The list is being used as was intended anyhow.
> 
> Also `tertiary` dye should be recoloring only the orle - tested it too. Secondary is usually plume. Primary is the helmet. If accessory dyeing colours the _entire_ thing that's probably a buuuuug?

Oh, I had no idea it'd hypothetically be this easy. Huh! Maybe I'll tinker with it and try this.

MortoSasye commented:
Hi. Darkest night seems to have vanished from the selection.

coffeecoffeecoffee commented:
> Hi. Darkest night seems to have vanished from the selection.

Lemme fix that!

coffeecoffeecoffee commented:
I have no idea why it would add the "balance" tag, but I've changed a few things.

1) Fixed "Darkest Night" #414143

2) The seamster/seamstress and tailors have been given a dyes trait. This allows THEM to access the color wheel! I was really enchanted by the idea of making this a unique player interaction thing to create business and flavor for tailors/seamsters ("Oh, you like this color? I got the tailor to do it for me!") and someone in dev chat brought up the idea of having a sort of "pantone magazine" that a Tailor or Seamster can give to a customer as a sort of in-character way for people to select and request an exact hex code. I'm IN LOVE with this idea, but I'd need help coding it as it's beyond my abilities.

3) Did you know the Naledi warscholars had their own unique color selection too? It's been unified with the overall color presets and they now have access to the whole breadth of colors, diversifying their choices massively. I'm especially in love with Wine Rouge on them.

chemistrymain3 commented:
Keep the color wheel. Don't pull out another dev's hard work over a tantrum you decided to have.

coffeecoffeecoffee commented:
> Keep the color wheel. Don't pull out another dev's hard work over a tantrum you decided to have.

I have, if you read the last post here - I've made it a feature of tailors and seamstresses.

radarseas commented:
hmm, I feel dyemaster would be interesting in the hands of merchants too. Sure they might step on the tailor's shoes sometimes, but i think the idea of a chance of them _selling custom dyed clothing at exorbitant prices in comparison would be fucking hilarious._

coffeecoffeecoffee commented:
don't PR now, working on a tailor's swatchbook as an alternative way for people to access the color wheel and the code for the dye bin is NOT working yet

coffeecoffeecoffee commented:
SO, now the tailors and seamsters spawn with a new book - the _Tailor's Swatchbook!_
![dreamseeker_8KueKzzgx1](https://github.com/user-attachments/assets/c0fab823-dfcd-4440-b522-b940dc0882cd)
Using this grants anybody access to the color wheel, so non-Dyer characters can pick out a color and use it. It has more or less the same luminosity restrictions as the normal color wheel option on the dye bin did. You don't need a tailor or seamster to make use of it - just slap a dye bin with it in their absence! I've mapped them into the Merchant's and the Seamster's, so they have one to give away or sell if they so wish. Or so you can swipe it. You naughty player, you.

Hopefully this is enough color wheel for everybody without just giving people full unfettered access at the very start of the round. Plus, gives tailors a little something more to do and their own cute little book. Hoping to add this to the merchant's goldface options so they can compete later.
